x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java

```diff
@@ -204,18 +204,23 @@ public static int copyToCacheFileAligned(IO fc, InputStream input, int fileChann
             if (bytesRead <= 0) {
                 break;
             }
-            if (buffer.hasRemaining()) {
-                // ensure that last write is aligned on 4k boundaries (= page size)
-                final int remainder = buffer.position() % PAGE_SIZE;
-                final int adjustment = remainder == 0 ? 0 : PAGE_SIZE - remainder;
-                buffer.position(buffer.position() + adjustment);
-            }
-            bytesCopied += positionalWrite(fc, fileChannelPos + bytesCopied, buffer);
+            bytesCopied += copyToCacheFileAligned(fc, fileChannelPos + bytesCopied, buffer);
             progressUpdater.accept(bytesCopied);
         }
         return bytesCopied;
     }
 
+    public static int copyToCacheFileAligned(IO fc, int fileChannelPos, ByteBuffer buffer) throws IOException {
```

Comment:
Can we add javadoc to this? Also, I'd prefer a different name, since it is not as much an overload of the original method (which uses the buffer as working area, this one uses it as real input). Perhaps:

```suggestion
    public static int copyBufferToCacheFileAligned(IO fc, int fileChannelPos, ByteBuffer buffer) throws IOException {
```